java/code/src/com/redhat/rhn/common/hibernate/AbstractConnectionManager.java

```diff
@@ -0,0 +1,334 @@
+        final SessionInfo info = threadSessionInfo();
+        if (info == null || info.getSession() == null) {
+            return;
+        }
+
+        final Transaction txn = info.getTransaction();
+        if (txn != null) {
+            txn.rollback();
+            info.setTransaction(null);
+        }
+    }
+
+    /**
+     * Returns Hibernate session stored in ThreadLocal storage. If not
+     * present, creates a new one and stores it in ThreadLocal; creating the
+     * session also begins a transaction implicitly.
+     * @return Session asked for
+     */
+    public Session getSession() {
+        if (!isInitialized()) {
+            initialize();
+        }
+
+        SessionInfo info = threadSessionInfo();
+        if (info == null || info.getSession() == null) {
+            try {
+                if (LOG.isDebugEnabled()) {
+                    LOG.debug("YYY Opening Hibernate Session");
+                }
+                info = new SessionInfo(sessionFactory.openSession());
+            }
+            catch (HibernateException e) {
+                throw new HibernateRuntimeException("couldn't open session", e);
+            }
+            sessionInfoThreadLocal.set(info);
+        }
+
+        // Automatically start a transaction
+        if (info.getTransaction() == null) {
+            info.setTransaction(info.getSession().beginTransaction());
+        }
+
+        return info.getSession();
+    }
+
+    /**
+     * Returns Hibernate session stored in ThreadLocal storage, if it exists
+     * @return Session a session
+     */
+    public Optional<Session> getSessionIfPresent() {
+        return Optional.ofNullable(threadSessionInfo()).map(SessionInfo::getSession);
+    }
+
+    /**
+     * Closes Hibernate Session stored in ThreadLocal storage.
+     */
+    public void closeSession() {
+        SessionInfo info = threadSessionInfo();
+        if (info == null) {
+            return;
+        }
+
+        Session session = info.getSession();
+        try {
+            Transaction txn = info.getTransaction();
+            if (txn != null && txn.getStatus().isNotOneOf(COMMITTED, ROLLED_BACK)) {
+                try {
+                    txn.commit();
+                }
+                catch (HibernateException e) {
+                    LOG.warn("Unable to commit transaction", e);
+                    txn.rollback();
+                }
+            }
+        }
+        catch (HibernateException e) {
+            LOG.error(e);
+        }
+        finally {
+            try {
+                if (session != null && session.isOpen()) {
+                    if (LOG.isDebugEnabled()) {
+                        LOG.debug("YYY Closing Hibernate Session");
+                    }
+                    session.close();
+                }
+            }
+            catch (HibernateException e) {
+                throw new HibernateRuntimeException("couldn't close session");
```

Comment:
This ```throw``` here can be tricky, because it's within a ```finally``` block, so it can happen when the code already threw an exception, and if it does, it hides to original exception. We indeed do log it above in the ```log.error()```, but it might be not that easy to figure out that the exceptions are connected. Not sure if we can change this though, because it will change the existing behaviour, even if this scenario should rarely happen.